src/main/scala/com/mongodb/spark/MongoSpark.scala

```diff
@@ -0,0 +1,510 @@
+   */
+  val defaultSource = "com.mongodb.spark.sql.DefaultSource"
+
+  /**
+   * Create a builder for configuring the [[MongoSpark]]
+   *
+   * @return a MongoSession Builder
+   */
+  def builder(): Builder = new Builder
+
+  /**
+   * Load data from MongoDB
+   *
+   * @param sc the Spark context containing the MongoDB connection configuration
+   * @return a MongoRDD
+   */
+  def load[D: ClassTag](sc: SparkContext)(implicit e: D DefaultsTo Document): MongoRDD[D] = load(sc, ReadConfig(sc))
+
+  /**
+   * Load data from MongoDB
+   *
+   * @param sc the Spark context containing the MongoDB connection configuration
+   * @return a MongoRDD
+   */
+  def load[D: ClassTag](sc: SparkContext, readConfig: ReadConfig)(implicit e: D DefaultsTo Document): MongoRDD[D] =
+    builder().sparkContext(sc).readConfig(readConfig).build().toRDD[D]()
+
+  /**
+   * Load data from MongoDB
+   *
+   * @param sqlContext the SQLContext containing the MongoDB connection configuration
+   * @tparam D The optional class defining the schema for the data
+   * @return a MongoRDD
+   */
+  def load[D <: Product: TypeTag](sqlContext: SQLContext): DataFrame =
+    load[D](sqlContext, ReadConfig(sqlContext.sparkContext))
+
+  /**
+   * Load data from MongoDB
+   *
+   * @param sqlContext the SQLContext containing the MongoDB connection configuration
+   * @tparam D The optional class defining the schema for the data
+   * @return a MongoRDD
+   */
+  def load[D <: Product: TypeTag](sqlContext: SQLContext, readConfig: ReadConfig): DataFrame =
+    builder().sqlContext(sqlContext).readConfig(readConfig).build().toDF[D]()
+
+  /**
+   * Load data from MongoDB
+   *
+   * @param sqlContext the SQL context containing the MongoDB connection configuration
+   * @param clazz   the class of the data contained in the RDD
+   * @tparam D The bean class defining the schema for the data
+   * @return a MongoRDD
+   */
+  def load[D](sqlContext: SQLContext, readConfig: ReadConfig, clazz: Class[D]): DataFrame =
+    builder().sqlContext(sqlContext).readConfig(readConfig).build().toDF(clazz)
+
+  /**
+   * Save data to MongoDB
+   *
+   * Uses the `SparkConf` for the database and collection information
+   * Requires a codec for the data type
+   *
+   * @param rdd the RDD data to save to MongoDB
+   * @tparam D the type of the data in the RDD
+   */
+  def save[D: ClassTag](rdd: RDD[D]): Unit = save(rdd, WriteConfig(rdd.sparkContext))
+
+  /**
+   * Save data to MongoDB
+   *
+   * @param rdd the RDD data to save to MongoDB
+   * @param writeConfig the writeConfig
+   * @tparam D the type of the data in the RDD
+   */
+  def save[D: ClassTag](rdd: RDD[D], writeConfig: WriteConfig): Unit = {
+    val mongoConnector = MongoConnector(writeConfig.asOptions)
+    rdd.foreachPartition(iter => if (iter.nonEmpty) {
+      mongoConnector.withCollectionDo(writeConfig, { collection: MongoCollection[D] =>
+        collection.insertMany(iter.toList.asJava)
```

Comment:
I know this is just refactored code and hasn't really changed, but I realizing looking at it that if iter is large # of docs calling toList on it could blow up memory